File: src/lib/crowd.ts
export const CHAT_OPTIONS = [
  { id: "love", label: "Love this" },
  { id: "beautiful", label: "So beautiful" },
  { id: "wow", label: "Wow" },
  { id: "withyou", label: "We're with you" },
  { id: "cheers", label: "Cheers" },
  { id: "clap", label: "Clapping" },
] as const;

export const FEEDBACK_GROUPS = [
  {
    kind: "picture",
    title: "Picture",
    options: [
      { id: "clear", label: "Clear" },
      { id: "soft", label: "A bit soft" },
      { id: "stuck", label: "Freezing" },
    ],
  },
  {
    kind: "sound",
    title: "Sound",
    options: [
      { id: "clear", label: "Clear" },
      { id: "quiet", label: "Too quiet" },
      { id: "none", label: "No sound" },
    ],
  },
  {
    kind: "moment",
    title: "The event",
    options: [
      { id: "great", label: "Great" },
      { id: "moving", label: "Moving" },
      { id: "issues", label: "Having issues" },
    ],
  },
] as const;

export type ChatOptionId = (typeof CHAT_OPTIONS)[number]["id"];
export type FeedbackKind = (typeof FEEDBACK_GROUPS)[number]["kind"];

export type ChatLine = {
  from: string;
  id: ChatOptionId;
  label: string;
  at: number;
};

const chatById = new Map(CHAT_OPTIONS.map((o) => [o.id, o.label]));
const feedbackOk = new Map<string, Set<string>>(
  FEEDBACK_GROUPS.map((g) => [g.kind, new Set(g.options.map((o) => o.id))]),
);

export function chatLabel(id: string) {
  return chatById.get(id as ChatOptionId) ?? null;
}

export function isChatId(id: string): id is ChatOptionId {
  return chatById.has(id as ChatOptionId);
}

export function isFeedbackChoice(kind: string, choice: string) {
  return feedbackOk.get(kind)?.has(choice) ?? false;
}

File: android/core/src/main/kotlin/com/eventview/core/Crowd.kt
package com.eventview.core

data class ChatOption(val id: String, val label: String)

data class ChatLine(
    val from: String,
    val id: String,
    val label: String,
    val at: Long,
)

data class FeedbackOption(val id: String, val label: String)

data class FeedbackGroup(
    val kind: String,
    val title: String,
    val options: List<FeedbackOption>,
)

/** Mirrors `src/lib/crowd.ts` — chat rides LiveKit data messages `{ t, v }`. */
object Crowd {
    val chatOptions: List<ChatOption> = listOf(
        ChatOption("love", "Love this"),
        ChatOption("beautiful", "So beautiful"),
        ChatOption("wow", "Wow"),
        ChatOption("withyou", "We're with you"),
        ChatOption("cheers", "Cheers"),
        ChatOption("clap", "Clapping"),
    )

    val feedbackGroups: List<FeedbackGroup> = listOf(
        FeedbackGroup(
            kind = "picture",
            title = "Picture",
            options = listOf(
                FeedbackOption("clear", "Clear"),
                FeedbackOption("soft", "A bit soft"),
                FeedbackOption("stuck", "Freezing"),
            ),
        ),
        FeedbackGroup(
            kind = "sound",
            title = "Sound",
            options = listOf(
                FeedbackOption("clear", "Clear"),
                FeedbackOption("quiet", "Too quiet"),
                FeedbackOption("none", "No sound"),
            ),
        ),
        FeedbackGroup(
            kind = "moment",
            title = "The stream",
            options = listOf(
                FeedbackOption("great", "Great"),
                FeedbackOption("moving", "Moving"),
                FeedbackOption("issues", "Having issues"),
            ),
        ),
    )

    private val chatById = chatOptions.associate { it.id to it.label }
    private val feedbackOk = feedbackGroups.associate { group ->
        group.kind to group.options.map { it.id }.toSet()
    }

    fun chatLabel(id: String): String? = chatById[id]

    fun isChatId(id: String): Boolean = chatById.containsKey(id)

    fun isFeedbackChoice(kind: String, choice: String): Boolean {
        return feedbackOk[kind]?.contains(choice) == true
    }

    fun isPictureTrouble(choice: String): Boolean = choice == "soft" || choice == "stuck"

    fun isSoundTrouble(choice: String): Boolean = choice == "quiet" || choice == "none"
}

data class LiveDataMessage(val t: String, val v: String)

object LiveData {
    fun encode(type: String, value: String): String = """{"t":"$type","v":"$value"}"""

    fun parse(raw: String): LiveDataMessage? {
        val t = Regex("\"t\"\\s*:\\s*\"([^\"]+)\"").find(raw)?.groupValues?.getOrNull(1) ?: return null
        val v = Regex("\"v\"\\s*:\\s*\"([^\"]+)\"").find(raw)?.groupValues?.getOrNull(1) ?: return null
        return LiveDataMessage(t, v)
    }
}
